docs/coding-guidelines/vectorization-guidelines.md

```diff
@@ -0,0 +1,1091 @@
+}
+```
+
+Benchmark all available solutions, and choose the one that is the best for us.
+
+```ini
+BenchmarkDotNet=v0.13.5, OS=Windows 11 (10.0.22621.1413/22H2/2022Update/SunValley2)
+AMD Ryzen Threadripper PRO 3945WX 12-Cores, 1 CPU, 24 logical and 12 physical cores
+.NET SDK=8.0.100-alpha.1.22558.1
+  [Host]   : .NET 7.0.4 (7.0.423.11508), X64 RyuJIT AVX2
+```
+
+```
+|                     Method | Size |      Mean | Ratio | Code Size |
+|--------------------------- |----- |----------:|------:|----------:|
+|                     Scalar | 1024 | 252.13 ns |  1.00 |      69 B |
+|             GreaterThanAny | 1024 |  32.49 ns |  0.13 |     178 B |
+|                LessThanAny | 1024 |  29.33 ns |  0.12 |     146 B |
+|                        And | 1024 |  26.13 ns |  0.10 |     138 B |
+|                      TestZ | 1024 |  27.26 ns |  0.11 |     129 B |
+| ExtractMostSignificantBits | 1024 |  27.33 ns |  0.11 |     141 B |
+```
+
+Even such a simple problem can be solved in at least 5 different ways. Using sophisticated hardware-specific instructions does not always provide the best performance, so **with the new `Vector128` and `Vector256` APIs we don't need to become assembly language experts to write fast, vectorized code**.
+
+## Toolchain
+
+`Vector128`, `Vector128<T>`, `Vector256` and `Vector256<T>` expose a LOT of APIs. We are constrained by time, so we won't describe all of them with examples. Instead, we have grouped them into categories to give you an overview of their capabilities. It's not required to remember what each of these methods is doing, but it's important to remember what kind of operations they allow for and check the details when needed.
+
+**Note:** all of these methods have "software fallbacks", which are executed when they cannot be vectorized on given platform.
+
+### Creation
+
+Each of the vector types provides a `Create` method that accepts a single value and returns a vector with all elements initialized to this value.
+
+```cs
+public static Vector128<T> Create<T>(T value) where T : struct;
+```
+
+`CreateScalar` initializes first element to the specified value, and the remaining elements to zero.
+
+```cs
+public static Vector128<int> CreateScalar(int value);
+```
+
+`CreateScalarUnsafe` is similar, but the remaining elements are left uninitialized. It's dangerous!
+
+
+We also have an overload that allows for specifying every value in given vector:
+
+```cs
+public static Vector128<short> Create(short e0, short e1, short e2, short e3, short e4, short e5, short e6, short e7)
+```
+
+And last but not least we have a `Create` overload which accepts a buffer. It creates a vector with its elements set to the first `VectorXYZ<T>.Count` elements of the buffer. It's not recommended to use it in a loop, where `Load` methods should be used instead (for performance).
+
+```cs
+public static Vector128<T> Create<T>(ReadOnlySpan<T> values) where T : struct
+```
```

Comment:
The span overloads require slicing as well as corresponding bounds checks and other things each iteration. This can sometimes be hoisted by the JIT, but it isn't always trivial to do so and the overhead can add up.